#Author : keime
#Date : 2011/09

###############################################################################################
#This function generates 2 graphs by using statistics calculated by FastQC
#1. boxplot of quality score in each sequencing cycle
#2. % of each nt in each sequencing cycle
#Arguments required :
#file : fastqc_data.txt generated by FastQC
#title : name of the corresponding png file
#dir : directory where to stores png files
#seqlengthline : line of sequence line information
#qualityscoreline : first line of quality score information
#seqcontentline :  first line of sequence content information
#Nseqcontentline : fisrt line of N content information
#################################################################################################

qualityanalysis = function(file, title, dir, seqlengthline, qualityscoreline, seqcontentline, Nseqcontentline){
	
	print(paste("file",file,"title",title,"dir",dir))
	
	#sequence length
	################
	data = read.table(file, skip=seqlengthline-1, nrow=1, h=F)
	seqlength = data$V3 
	
	#total nb of sequences
	######################
	#data = read.table(file, skip=seqlengthline-3, nrow=1, h=F) #to be compatible with FastQC v0.10.0 and higher -3 because line "filtered sequences" (for previous version of FastQC use -2)
	#seqnb = data$V3
	#not useful anymore because the percentage are directly available in the fastqc_data.txt file (no more need to calculate these %)
	
	#boxplot of quality score = f(position)
	#######################################
	#read corresponding data
	data=read.table(file, skip=qualityscoreline, nrows=seqlength, h=T, sep="\t", comment.char="")
	
	#open output pdf file
	pngfile=paste(paste(dir,title,sep="/"),".png",sep="")
	png(pngfile, width=1000, height=500)
	par(mfrow=c(1,2))
	
	#draw boxplot
	stats=t(data.frame(data$X10th.Percentile,data$Lower.Quartile,data$Median,data$Upper.Quartile,data$X90th.Percentile))
	bx = list(stats=stats,names=data$X.Base)
	bxp(bx,xlab="position",ylab="quality score",ylim=c(0,40),main="Boxplots of the phred quality scores\n along the sequencing cycles", ylim=c(0,40))
	
	
	#% of each base in each cycle
	#############################
	#percentage of A,C,G and T at each position 
	#Caution, since FastQC 0.9.5 the percentage is indicated in text file, ie this is no longer the number of each base at each position
	#but the percentage is calculated only relative to A, C, G and T bases (this does not include N bases)
	data = read.table(file, skip=seqcontentline, nrows=seqlength, h=T, sep="\t", comment.char="", row.names=1)
	Apercent = data$A
	Cpercent = data$C
	Gpercent = data$G
	Tpercent = data$T
	
	#% of N at each position
	data = read.table(file, skip=Nseqcontentline, nrows=seqlength, h=T, sep="\t", comment.char="", row.names=1)
	Npercent = data$N.Count #caution, this is called Count in fastqc, but it is a percentage
	
	#percentage of each base at each position
	#we calculate the real percentage, ie including all bases 
	Apercentok = Apercent * (100-Npercent) / 100
	Cpercentok = Cpercent * (100-Npercent) / 100
	Gpercentok = Gpercent * (100-Npercent) / 100
	Tpercentok = Tpercent * (100-Npercent) / 100
	allpercent = t(data.frame(Apercentok, Cpercentok, Gpercentok, Tpercentok, Npercent))
	rownames(allpercent) = c("A", "C", "G", "T","N")
	
	#draw barplot
	barplot(allpercent,col=2:6,legend.text=rownames(allpercent),args.legend = c(x=seqlength*1.34,y=23.5),
			xlab="position",ylab="% of each base",names.arg=1:dim(allpercent)[2],main="Repartition of the different nucleotides\n along the sequencing cycles")
	
	dev.off()
	
}

